core/include/core/G3Timestream.h

```diff
@@ -143,31 +144,107 @@ class G3Timestream : public G3FrameObject {
 		TS_INT32,
 		TS_INT64
 	} data_type_;
+	
+	template<typename T>
+	struct TimeStreamTypeResolver{
+		static_assert(sizeof(T)!=sizeof(T), "Unsupported datatype for G3Timestream");
+	};
 };
 
 G3_POINTERS(G3Timestream);
 
+template<>
+struct G3Timestream::TimeStreamTypeResolver<double>{
+	constexpr static auto type_tag=G3Timestream::TS_DOUBLE;
+};
+
+template<>
+struct G3Timestream::TimeStreamTypeResolver<float>{
+	constexpr static auto type_tag=G3Timestream::TS_FLOAT;
+};
+
+template<>
+struct G3Timestream::TimeStreamTypeResolver<int32_t>{
+	constexpr static auto type_tag=G3Timestream::TS_INT32;
+};
+
+template<>
+struct G3Timestream::TimeStreamTypeResolver<int64_t>{
+	constexpr static auto type_tag=G3Timestream::TS_INT64;
+};
+
 class G3TimestreamMap : public G3FrameObject,
     public std::map<std::string, G3TimestreamPtr> {
 public:
+	G3TimestreamMap(){}
+	G3TimestreamMap(const G3TimestreamMap& other):std::map<std::string, G3TimestreamPtr>(other){}
+	G3TimestreamMap(G3TimestreamMap&& other):std::map<std::string, G3TimestreamPtr>(std::move(other)){}
+
 	// Return true if all timestreams start and end at the same time and
 	// contain the same number of samples.
 	bool CheckAlignment() const;
 
 	// Return statistics for contained timestreams. Results undefined if
 	// CheckAlignment is false.
 	G3Time GetStartTime() const;
+	void SetStartTime(G3Time start);
 	G3Time GetStopTime() const;
+	void SetStopTime(G3Time stop);
 	double GetSampleRate() const;
 	size_t NSamples() const;
 	G3Timestream::TimestreamUnits GetUnits() const;
+	void SetUnits(G3Timestream::TimestreamUnits units);
+	/// FLAC compression levels range from 0-9. 0 means do not use FLAC.
+	void SetFLACCompression(int compression_level);
 
 	// Compact underlying data storage into a contiguous 2D block.
 	// This invalidates any references to data inside any member
 	// timestreams (though not the timestream objects themselves)
 	// and requires timestream alignment (throws exception if
 	// CheckAlignment is false).
 	void Compactify();
+	
+	/// Construct a map whose underlying data storage is a contiguous 2D block.
+	/// \param keys the timestream keys for which the map should be constructed
+	/// \param start the start time which will be shared by all time streams
+	/// \param stop the stop time which will be shared by all time streams
+	/// \param n_samples the number of samples each timestream will have
+	/// \pre keys must be in sorted order
+	template<typename SampleType>
+	static G3TimestreamMap MakeCompact(const std::vector<std::string>& keys, G3Time start, G3Time stop, std::size_t n_samples){
```

Comment:
Do we need / want python bindings for these methods?

Reply:
As Nathan has mentioned, the idea in large part was to enable doing what could already be done from python (via `timestream_from_iterable`, I think), but directly in C++ without numpy. As such, I think python mostly doesn't need this. However, it looks like the features may not be entirely synchronized, as it looks like python can set the units as part of the same operation, while I included start and stop time instead. I could be useful to make both interfaces support all of those.

Reply:
Agreed on that last point -- both interfaces should support any of the configurable options, including compression level.

Reply:
Looking more carefully, I think the corresponding function is actually `G3TimestreamMap_from_numpy()`, which has `keys`, `data`, `start`, `stop`, `units`, and `copy_data`. I can add `units` to `MakeCompact` and, `compression_level` to both. Does it seem worthwhile to implement `copy_data` for `MakeCompact`? I can do it, but I personally don't ever want to use it, and I'm not sure if anyone else would either.

Reply:
I'm fine with not implementing copy_data in C++!

Reply:
This should now be cleaned up.